plugins/calibre.koplugin/parser.lua

```diff
@@ -1,90 +1,93 @@
+    if array_fields[field] and t[field] then
+        table.insert(t[field], v)
     else
-        return replaceHexChars(str, key:len() + 10, key == "series_index" and 2 or 3)
+        t[field] = v
+        field = nil
     end
 end
 
-local jsonStr = getmetatable("")
-jsonStr.__index["equals"] = isEqual
-jsonStr.__index["value"] = getValue
-
-
-local parser = {}
-
--- read metadata from file, line by line, and keep just the data we need
-function parser.parseFile(file)
-    assert(type(file) == "string", "wrong type (expected a string")
-    local f, err = io.open(file, "rb")
-    if not f then
-        return nil, string.format("error parsing %s: %s", file, err)
-    end
-    f:close()
-    local add = function(t, line)
-        if type(t) ~= "table" or type(line) ~= "string" then
-            return {}
+local depth = 0
+local result = {}
+local sax = {
+    startobject = function()
+        depth = depth + 1
+    end,
+    endobject = function()
+        if depth == 1 then
+            table.insert(result, t)
+            t = {}
         end
-        line = replaceHexChars(line, 8, 3)
-        table.insert(t, #t + 1, line)
-        return t
-    end
-    local books, book = {}, {}
-    local is_author, is_tag = false, false
-    for line in io.lines(file) do
-        if line == "  }, " or line == "  }" then
-            if type(book) == "table" then
-                table.insert(books, #books + 1, book)
-            end
-            book = {}
-        elseif line == "    \"authors\": [" then
-            is_author = true
-        elseif line == "    \"tags\": [" then
-            is_tag = true
-        elseif line == "    ], " or line == "    ]" then
-            is_author, is_tag = false, false
-        else
-            for _, key in ipairs({"title", "uuid", "lpath", "size",
-                "last_modified", "series", "series_index"})
-            do
-                if line:equals(key) then
-                    book[key] = line:value(key)
-                    break
-                end
-            end
+        depth = depth - 1
+    end,
+    startarray = function()
+        if array_fields[field] then
+            t[field] = {}
         end
-        if is_author then
-            book.authors = add(book.authors, line)
-        elseif is_tag then
-            book.tags = add(book.tags, line)
+    end,
+    endarray = function()
+        if field then
+            field = nil
         end
-    end
-    return books
+    end,
+    key = function(s)
+        if required_fields[s] then
+            field = s
+        end
+    end,
+    string = function(s)
+        if field then
+            append(s)
+        end
+    end,
+    number = function(n)
+        if field then
+            append(n)
+        end
+    end,
+    boolean = function(b)
+        if field then
+            append(b)
+        end
+    end,
+    null = function()
+        if field then
+            append(nil)
+        end
+    end,
```

Comment:
It also doesn't make much of a difference, but I figure we can also get rid of the `null` handler, as `t[field] = nil` is effectively a NOP.

Reply:
(And LunaJSON will use an actual nop if a handler isn't set in the sax table).